### PR TITLE
Allow configuring passes for optimizations on EXPLAIN PLAN

### LINKED ISSUE #48711: Add a setting `optimizations` to `EXPLAIN PLAN`
For `EXPLAIN QUERY TREE` there is a setting `passes` which limits the number of passes run on top of the query tree. [Example](https://pastila.nl/?00110568/ff794362f4c8d5bd18912f0037218091).

I want the same setting to limit the number of optimizations applied to the query plan.
Example:

```
EXPLAIN actions = 1, optimize = 1
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)

Query id: 45a4b3aa-269e-4fca-995c-ebb17d5026e2

┌─explain───────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + (Before ORDER BY + (Projection + (Before ORDER BY + (Projection + Before ORDER BY)))))) │
│ Actions: INPUT : 0 -> number UInt64 : 0                                                                           │
│          COLUMN Const(UInt8) -> 1 UInt8 : 1                                                                       │
│          COLUMN Const(UInt8) -> 1 UInt8 : 2                                                                       │
│          COLUMN Const(UInt8) -> 1 UInt8 : 3                                                                       │
│          FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 4                                         │
│          ALIAS plus(1, number) :: 4 -> x UInt64 : 0                                                               │
│          FUNCTION plus(1 :: 2, x :: 0) -> plus(1, x) UInt64 : 4                                                   │
│          ALIAS plus(1, x) :: 4 -> y UInt64 : 0                                                                    │
│          FUNCTION plus(1 :: 3, y :: 0) -> plus(1, y) UInt64 : 4                                                   │
│          ALIAS plus(1, y) :: 4 -> z UInt64 : 0                                                                    │
│ Positions: 0                                                                                                      │
│   ReadFromStorage (SystemNumbers)                                                                                 │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘


EXPLAIN actions = 1, optimize = 0
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)

Query id: 1481a5c6-730d-4c6b-b3b3-5a1aa48f9d6d

┌─explain─────────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                             │
│ Actions: INPUT : 0 -> plus(1, y) UInt64 : 0                                         │
│          ALIAS plus(1, y) :: 0 -> z UInt64 : 1                                      │
│ Positions: 1                                                                        │
│   Expression (Before ORDER BY)                                                      │
│   Actions: INPUT : 0 -> y UInt64 : 0                                                │
│            COLUMN Const(UInt8) -> 1 UInt8 : 1                                       │
│            FUNCTION plus(1 :: 1, y :: 0) -> plus(1, y) UInt64 : 2                   │
│   Positions: 2                                                                      │
│     Expression (Projection)                                                         │
│     Actions: INPUT : 0 -> plus(1, x) UInt64 : 0                                     │
│              ALIAS plus(1, x) :: 0 -> y UInt64 : 1                                  │
│     Positions: 1                                                                    │
│       Expression (Before ORDER BY)                                                  │
│       Actions: INPUT : 0 -> x UInt64 : 0                                            │
│                COLUMN Const(UInt8) -> 1 UInt8 : 1                                   │
│                FUNCTION plus(1 :: 1, x :: 0) -> plus(1, x) UInt64 : 2               │
│       Positions: 2                                                                  │
│         Expression (Projection)                                                     │
│         Actions: INPUT : 0 -> plus(1, number) UInt64 : 0                            │
│                  ALIAS plus(1, number) :: 0 -> x UInt64 : 1                         │
│         Positions: 1                                                                │
│           Expression (Before ORDER BY)                                              │
│           Actions: INPUT : 0 -> number UInt64 : 0                                   │
│                    COLUMN Const(UInt8) -> 1 UInt8 : 1                               │
│                    FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 2 │
│           Positions: 2                                                              │
│             ReadFromStorage (SystemNumbers)                                         │
└─────────────────────────────────────────────────────────────────────────────────────┘
```

Expected: something like
```
EXPLAIN actions = 1, optimizations = 1
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)
FORMAT TSV

Query id: 7aa46c44-e2c6-4d10-bf13-b874f2ae2fd6

Expression (Projection)
Actions: INPUT : 0 -> plus(1, y) UInt64 : 0
         ALIAS plus(1, y) :: 0 -> z UInt64 : 1
Positions: 1
  Expression (Before ORDER BY)
  Actions: INPUT : 0 -> y UInt64 : 0
           COLUMN Const(UInt8) -> 1 UInt8 : 1
           FUNCTION plus(1 :: 1, y :: 0) -> plus(1, y) UInt64 : 2
  Positions: 2
    Expression (Projection)
    Actions: INPUT : 0 -> plus(1, x) UInt64 : 0
             ALIAS plus(1, x) :: 0 -> y UInt64 : 1
    Positions: 1
      Expression (Before ORDER BY)
      Actions: INPUT : 0 -> x UInt64 : 0
               COLUMN Const(UInt8) -> 1 UInt8 : 1
               FUNCTION plus(1 :: 1, x :: 0) -> plus(1, x) UInt64 : 2
      Positions: 2
        Expression ((Projection + Before ORDER BY))
        Actions: INPUT : 0 -> number UInt64 : 0
                 COLUMN Const(UInt8) -> 1 UInt8 : 1
                 FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 2
                 ALIAS plus(1, number) :: 2 -> x UInt64 : 0
        Positions: 0
          ReadFromStorage (SystemNumbers)
```

### PR DESCRIPTION
**Pull request by GMartinez-Sisti**
Closes https://github.com/ClickHouse/ClickHouse/issues/48711.

### Changelog category (leave one):
- New Feature

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
Allow configuring passes for optimizations on EXPLAIN PLAN

### Examples
<details>
  <summary>EXPLAIN actions = 1, optimize = 1, passes = 0</summary>

```sh
EXPLAIN actions = 1, optimize = 1, passes = 0
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)

Query id: 93afd5bb-fb19-4382-9dad-e79647e9b85a

┌─explain─────────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                             │
│ Actions: INPUT : 0 -> plus(1, y) UInt64 : 0                                         │
│          ALIAS plus(1, y) :: 0 -> z UInt64 : 1                                      │
│ Positions: 1                                                                        │
│   Expression (Before ORDER BY)                                                      │
│   Actions: INPUT : 0 -> y UInt64 : 0                                                │
│            COLUMN Const(UInt8) -> 1 UInt8 : 1                                       │
│            FUNCTION plus(1 :: 1, y :: 0) -> plus(1, y) UInt64 : 2                   │
│   Positions: 2                                                                      │
│     Expression (Projection)                                                         │
│     Actions: INPUT : 0 -> plus(1, x) UInt64 : 0                                     │
│              ALIAS plus(1, x) :: 0 -> y UInt64 : 1                                  │
│     Positions: 1                                                                    │
│       Expression (Before ORDER BY)                                                  │
│       Actions: INPUT : 0 -> x UInt64 : 0                                            │
│                COLUMN Const(UInt8) -> 1 UInt8 : 1                                   │
│                FUNCTION plus(1 :: 1, x :: 0) -> plus(1, x) UInt64 : 2               │
│       Positions: 2                                                                  │
│         Expression (Projection)                                                     │
│         Actions: INPUT : 0 -> plus(1, number) UInt64 : 0                            │
│                  ALIAS plus(1, number) :: 0 -> x UInt64 : 1                         │
│         Positions: 1                                                                │
│           Expression (Before ORDER BY)                                              │
│           Actions: INPUT : 0 -> number UInt64 : 0                                   │
│                    COLUMN Const(UInt8) -> 1 UInt8 : 1                               │
│                    FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 2 │
│           Positions: 2                                                              │
│             ReadFromStorage (SystemNumbers)                                         │
└─────────────────────────────────────────────────────────────────────────────────────┘

28 rows in set. Elapsed: 0.024 sec.
```

</details>

<details>
  <summary>EXPLAIN actions = 1, optimize = 1, passes = 2</summary>

```sh
EXPLAIN actions = 1, optimize = 1, passes = 2
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)

Query id: afc95a76-2e79-4901-9841-9b135a5a9ee5

┌─explain─────────────────────────────────────────────────────────────────────────┐
│ Expression (Projection)                                                         │
│ Actions: INPUT : 0 -> plus(1, y) UInt64 : 0                                     │
│          ALIAS plus(1, y) :: 0 -> z UInt64 : 1                                  │
│ Positions: 1                                                                    │
│   Expression (Before ORDER BY)                                                  │
│   Actions: INPUT : 0 -> y UInt64 : 0                                            │
│            COLUMN Const(UInt8) -> 1 UInt8 : 1                                   │
│            FUNCTION plus(1 :: 1, y :: 0) -> plus(1, y) UInt64 : 2               │
│   Positions: 2                                                                  │
│     Expression (Projection)                                                     │
│     Actions: INPUT : 0 -> plus(1, x) UInt64 : 0                                 │
│              ALIAS plus(1, x) :: 0 -> y UInt64 : 1                              │
│     Positions: 1                                                                │
│       Expression ((Before ORDER BY + (Projection + Before ORDER BY)))           │
│       Actions: INPUT : 0 -> number UInt64 : 0                                   │
│                COLUMN Const(UInt8) -> 1 UInt8 : 1                               │
│                COLUMN Const(UInt8) -> 1 UInt8 : 2                               │
│                FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 3 │
│                ALIAS plus(1, number) :: 3 -> x UInt64 : 0                       │
│                FUNCTION plus(1 :: 2, x :: 0) -> plus(1, x) UInt64 : 3           │
│       Positions: 3                                                              │
│         ReadFromStorage (SystemNumbers)                                         │
└─────────────────────────────────────────────────────────────────────────────────┘

22 rows in set. Elapsed: 0.024 sec.
```

</details>

<details>
  <summary>EXPLAIN actions = 1, optimize = 1, passes = 10</summary>

```sh
EXPLAIN actions = 1, optimize = 1, passes = 10
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)

Query id: 0504051c-eb55-437c-aa37-d63ba760c077

┌─explain───────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + (Before ORDER BY + (Projection + (Before ORDER BY + (Projection + Before ORDER BY)))))) │
│ Actions: INPUT : 0 -> number UInt64 : 0                                                                           │
│          COLUMN Const(UInt8) -> 1 UInt8 : 1                                                                       │
│          COLUMN Const(UInt8) -> 1 UInt8 : 2                                                                       │
│          COLUMN Const(UInt8) -> 1 UInt8 : 3                                                                       │
│          FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 4                                         │
│          ALIAS plus(1, number) :: 4 -> x UInt64 : 0                                                               │
│          FUNCTION plus(1 :: 2, x :: 0) -> plus(1, x) UInt64 : 4                                                   │
│          ALIAS plus(1, x) :: 4 -> y UInt64 : 0                                                                    │
│          FUNCTION plus(1 :: 3, y :: 0) -> plus(1, y) UInt64 : 4                                                   │
│          ALIAS plus(1, y) :: 4 -> z UInt64 : 0                                                                    │
│ Positions: 0                                                                                                      │
│   ReadFromStorage (SystemNumbers)                                                                                 │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

13 rows in set. Elapsed: 0.025 sec.
```

</details>

<details>
  <summary>EXPLAIN actions = 1, optimize = 1, passes = 1000</summary>

```sh
EXPLAIN actions = 1, optimize = 1, passes = 1000
SELECT 1 + y AS z
FROM
(
    SELECT 1 + x AS y
    FROM
    (
        SELECT 1 + number AS x
        FROM numbers(2)
    )
)

Query id: 616ae9fc-9731-438f-8169-dc2ec0deffcd

┌─explain───────────────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Expression ((Projection + (Before ORDER BY + (Projection + (Before ORDER BY + (Projection + Before ORDER BY)))))) │
│ Actions: INPUT : 0 -> number UInt64 : 0                                                                           │
│          COLUMN Const(UInt8) -> 1 UInt8 : 1                                                                       │
│          COLUMN Const(UInt8) -> 1 UInt8 : 2                                                                       │
│          COLUMN Const(UInt8) -> 1 UInt8 : 3                                                                       │
│          FUNCTION plus(1 :: 1, number :: 0) -> plus(1, number) UInt64 : 4                                         │
│          ALIAS plus(1, number) :: 4 -> x UInt64 : 0                                                               │
│          FUNCTION plus(1 :: 2, x :: 0) -> plus(1, x) UInt64 : 4                                                   │
│          ALIAS plus(1, x) :: 4 -> y UInt64 : 0                                                                    │
│          FUNCTION plus(1 :: 3, y :: 0) -> plus(1, y) UInt64 : 4                                                   │
│          ALIAS plus(1, y) :: 4 -> z UInt64 : 0                                                                    │
│ Positions: 0                                                                                                      │
│   ReadFromStorage (SystemNumbers)                                                                                 │
└───────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘

13 rows in set. Elapsed: 0.024 sec.
```

</details>

### CODE REVIEW

File: src/Processors/QueryPlan/QueryPlan.h [hunk 1]
```diff
@@ -58,8 +58,7 @@ class QueryPlan
     void optimize(const QueryPlanOptimizationSettings & optimization_settings);
 
     QueryPipelineBuilderPtr buildQueryPipeline(
-        const QueryPlanOptimizationSettings & optimization_settings,
-        const BuildQueryPipelineSettings & build_pipeline_settings);
+        const QueryPlanOptimizationSettings & optimization_settings, const BuildQueryPipelineSettings & build_pipeline_settings);
```
Comment (novikd) [inline src/Processors/QueryPlan/QueryPlan.h:61]: Let's avoid pointless reformatting in all files.

Comment (GMartinez-Sisti) [reply src/Processors/QueryPlan/QueryPlan.h]: Removed the linter commit. Had to force push.

**Comment by robot-clickhouse** [comment]
*This is an automated comment for commit 283bd18e7d9d7c1be5788619b8d0eb12529585c8 with description of existing statuses. It's updated for the latest CI running*

[❌ Click here](https://s3.amazonaws.com/clickhouse-test-reports/48900/283bd18e7d9d7c1be5788619b8d0eb12529585c8/ci_running.html) to open a full report in a separate page

<table>
<thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody>
<tr><td>Docs Check</td><td>Builds and tests the documentation</td><td>❌ failure</td></tr>
<tbody>
</table>

**Comment by GMartinez-Sisti** [comment]
The affected files have changed too much for me to fix this now. Will revisit later. Closing.

**Pull request closed without merging**